### PR TITLE
Fourier transform

### PR DESCRIPTION
**Pull request by laurgao**
Adds the functionality to use fourier transforms to analyze landscapes of cost functions. Also includes a jupyter notebook tutorial.

### CODE REVIEW

**Comment by MSRudolph** [comment]
Hey, thanks @laurgao! This is looking awesome. 
I'm wondering, for the purpose of composability, if the `fourier` methods should take in previous scan results, rather than performing a scan and then applying the fourier transform. It feels like it should be something to apply to your scan, rather than running a completely new one. Of course there may be good reasons to do so.

File: src/orqviz/fourier.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+            Defaults to None.
+        plot_kwargs: kwargs for plotting with matplotlib.pyplot.pcolormesh
+            (plt.pcolormesh)
+
+    Note: Fourier coefficients are complex numbers. However, due to the inability to
+    plot complex numbers on real axes, the magnitudes of the coefficients are plotted.
+    This means that phase has no influence on the visual output of the plot.
+    """
+    plottable_result = np.abs(result.values)
+    Nx = result.values.shape[1]
```
Comment (mstechly) [inline src/orqviz/fourier.py:106]: It's a little bit weird to use `Nx` as a variable name – I'd rather go with `n_x`, seems more pythonish to me.

File: src/orqviz/fourier.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        max_freq_y = min(Ny // 2, max_freq_x or np.inf)
+    if max_freq_x is None:
+        max_freq_x = Nx - 1
+
+    # normalize frequencies for range bigger than 1 period of 2pi
+    # (note that this is different from normalizing the coefficient magnitudes)
+    norm_x = (result.end_points_x[1] - result.end_points_x[0]) / (2 * np.pi)
+    norm_y = (result.end_points_y[1] - result.end_points_y[0]) / (2 * np.pi)
+    if max_freq_x > (Nx - 1) / norm_x:
+        warnings.warn(
```
Comment (mstechly) [inline src/orqviz/fourier.py:118]: There's a lot of cases here in if statements which are not covered by tests, could please add some tests for these?

File: src/orqviz/fourier.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+    The returned array is of size 2 * res + 1 in the y-direction and res + 1 in the
+    x-direction.
+    """
+    cy = (result.shape[0] - 1) // 2  # center
+    return result[cy - res_y : cy + res_y + 1, 0 : res_x + 1]
+    # Note this always makes result.shape[0] (y resolution) odd
+
+
+def _swap(result: np.ndarray) -> np.ndarray:
```
Comment (mstechly) [inline src/orqviz/fourier.py:183]: Could we rename it to something a little bit more verbose?

File: src/orqviz/fourier.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    """Swaps the result format from the format of the output of np.fft.rfft2 to the
+    format of the array used for plotting (where frequencies are lined up from least
+    to greatest going left to right).
+    """
+    Ny = result.shape[0]
+    return np.append(result[Ny // 2 + 1 :], result[0 : Ny // 2 + 1], axis=0)
+
+
+def _iswap(result: np.ndarray) -> np.ndarray:
```
Comment (mstechly) [inline src/orqviz/fourier.py:192]: Same here :)

**Review by mstechly — CHANGES_REQUESTED** [review]

**Comment by mstechly** [comment]
+ I agree with @MSRudolph's comment! 
I think we should have two methods:
One that allows to run fourier based on a scan that has been provided and the other one, which just runs fourier (as is).
This will give users two interfaces to interact with fourier, which I think would come in handy.

And I think the latter one will just use the first one under the hood, so it shouldn't be much code duplication :)

File: src/orqviz/__init__.py [hunk 5]
```diff
@@ -1,6 +1,7 @@
 from . import (
     aliases,
     elastic_band,
+    fourier,
```
[the remark this answers is not in the document]
Comment (laurgao) [reply src/orqviz/__init__.py:4]: That makes sense. But I do wonder if we don't plan to expand the spectral analysis to beyond fourier later, the name "fourier" might be more descriptive?

**Comment by laurgao** [comment]
Thank you for the comments @MSRudolph  and @mstechly !

File: src/orqviz/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mstechly) [reply src/orqviz/__init__.py:4]: Fair point 👌

**Review by mstechly — APPROVED** [review]

**Pull request merged**